### PR TITLE
[TFLite] Move AVERAGE_POOL_3D and MAX_POOL_3D to builtin ops

### PR DESCRIPTION
**Pull request by Tessil**
Hi,

This PR moves the custom `AvgPool3D` and `MaxPool3D` ops to bulitin `AVERAGE_POOL_3D` and `MAX_POOL_3D`  ops (`AVG` is changed to `AVERAGE` for coherence with `AVERAGE_POOL_2D`).

As discussed previously @jianlijianli we tried to keep the code of the custom ops for backward compatibility, notably in [tensorflow/lite/kernels/BUILD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/BUILD#L785) and [tensorflow/lite/kernels/custom_ops_register.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/custom_ops_register.h#L22). Note though that some of the code in [tensorflow/lite/kernels/pooling3d.cc](https://github.com/tensorflow/tensorflow/pull/53206/files#diff-5591b2bdc4d02753edd1722677e7f3e3645a2531a411c904665bb94f3870ec16) for reading the flatbuffer was removed which could create some backward incompatibilities. We could try to keep it in some way if needed but as the custom ops version of these kernels were not compiled into the official binary I'm more in favour of removing this old code if possible along with the one in `tensorflow/lite/kernels/BUILD` and `tensorflow/lite/kernels/custom_ops_register.h`.

Thibaut

### CODE REVIEW

**Comment by Tessil** [comment]
@miaout17 As discussed, please see https://arxiv.org/abs/2007.13224 for a network using MaxPool3D and https://keras.io/examples/vision/3D_image_classification/ example that uses it. It'd be nice to be able to convert the Keras model described in the example to TFLite.

Thanks,
Thibaut

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned Thanks, I fixed the conflicts

**Comment by miaout17** [comment]
Sorry for the wait. I think it looks good to add the 3D pooling ops as builtin. I'm looking for someone to review the code.

File: tensorflow/lite/kernels/pooling3d.cc
Comment (sirakiin) [inline tensorflow/lite/kernels/pooling3d.cc:233]: Shall we also keep Register_MAX_POOL_3D?

**Review by sirakiin — CHANGES_REQUESTED** [review]
Overall looks good. Thanks for the PR.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td]: Done

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td]: Done

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (Tessil) [reply tensorflow/lite/kernels/internal/reference/pooling3d.h]: PR was created in 2021, updated the copyright to 2022.

File: tensorflow/lite/kernels/internal/types.h
Comment (Tessil) [reply tensorflow/lite/kernels/internal/types.h]: Done

File: tensorflow/lite/tools/versioning/runtime_version.cc
Comment (Tessil) [reply tensorflow/lite/tools/versioning/runtime_version.cc]: Done

File: tensorflow/lite/kernels/pooling3d_test.cc
Comment (Tessil) [reply tensorflow/lite/kernels/pooling3d_test.cc]: Done

File: tensorflow/lite/kernels/pooling3d.cc
Comment (Tessil) [reply tensorflow/lite/kernels/pooling3d.cc]: Effectively, not sure why I didn't initially. Added the implementation for `custom::Register_MAX_POOL_3D`.

File: tensorflow/lite/testing/op_tests/pool3d.py
Comment (Tessil) [reply tensorflow/lite/testing/op_tests/pool3d.py:93]: Yes, the initial tests have 6 expected failures and we run the same tests for both int8 and int16, hence 18 expected failures.

**Comment by Tessil** [comment]
Thank you for the review.

Dusted off the PR a bit, fixed the merge conflicts and addressed the comments. Only some MLIR tests still need to be added which I will do when I have the occasion.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:3772]: Done

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td:5698]: Done

File: tensorflow/compiler/mlir/lite/transforms/legalize_patterns.td
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/transforms/legalize_patterns.td:583]: Done

**Comment by Tessil** [comment]
@sirakiin I added the MLIR tests, if you could check if everything looks good. Thanks

**Review by sirakiin — APPROVED** [review]
Looks good, many thanks for the contribution!

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned @sirakiin Thanks, I fixed the conflict. The PR will need re-approval.

**Review by sirakiin — APPROVED** [review]

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned Thanks, I fixed the conflicts @sirakiin It'll need re-approval. Thanks!

**Comment by abattery** [comment]
@alankelly any opinions on this operator addition?

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned Thanks, I fixed the conflicts.

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
@Tessil Any update on this PR? Please. Thank you!

**Comment by Tessil** [comment]
@gbaned Thanks, I fixed the conflict.

**Comment by gbaned** [comment]
@Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
Thanks, solved them.

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (alankelly) [inline tensorflow/lite/kernels/internal/reference/pooling3d.h:94]: Can you replace Offset with the partial calculation of the input index in each of the nested loops? Offset has terrible performance and I am working on getting rid of it.

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (alankelly) [inline tensorflow/lite/kernels/internal/reference/pooling3d.h:105]: Same

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (alankelly) [inline tensorflow/lite/kernels/internal/reference/pooling3d.h:174]: Same

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (alankelly) [inline tensorflow/lite/kernels/internal/reference/pooling3d.h:181]: Same

File: tensorflow/lite/tools/versioning/runtime_version.cc
Comment (alankelly) [inline tensorflow/lite/tools/versioning/runtime_version.cc:396]: 2.12.0

**Review by alankelly — COMMENTED** [review]
I'm OK with the kernel changes. Adding @arfaian for the converter changes

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (Tessil) [reply tensorflow/lite/kernels/internal/reference/pooling3d.h:94]: I moved the implementations from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/pooling3d.cc to their own `pooling3d.h` file. 

As the PR has been open since more than one year and it's an important op to have for us, can we do it in two steps to avoid complexifying this PR more than it's currently? Once merged, I can create an optimisation PR to remove the `Offset` usage if it's alright to you.

Thanks

File: tensorflow/lite/tools/versioning/runtime_version.cc
Comment (Tessil) [reply tensorflow/lite/tools/versioning/runtime_version.cc:396]: Changed it, thanks.

File: tensorflow/lite/kernels/internal/reference/pooling3d.h
Comment (alankelly) [reply tensorflow/lite/kernels/internal/reference/pooling3d.h:94]: OK for me. Thanks

File: tensorflow/lite/tools/versioning/runtime_version.cc
Comment (alankelly) [reply tensorflow/lite/tools/versioning/runtime_version.cc]: Thanks

**Review by alankelly — APPROVED** [review]

File: tensorflow/compiler/mlir/lite/utils/validators.cc
Comment (jpienaar) [inline tensorflow/compiler/mlir/lite/utils/validators.cc:99]: What can be it be except an IntegerAttr? I ask as in line 96 there is assumption that this is only used where ArrayAttr is already verified to be the attribute type and the function name says it should only be used for int lists. So seems like one assumption is checked while the other is assumed (and cast will just assert in debug mode).

File: tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir
Comment (jpienaar) [inline tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir:140]: I'm assuming you are doing this just to avoid dead code elimination? If so, how about using tf.AddN with a comment.

File: tensorflow/compiler/mlir/lite/tests/ops.mlir
Comment (jpienaar) [inline tensorflow/compiler/mlir/lite/tests/ops.mlir:683]: This is the "old" form, you can just roll the bb args into the function signature as you did in the other test.

**Review by jpienaar — COMMENTED** [review]
Moving to builtin is mostly question for TFLite team and seems they are happy with it. Gave the converter parts and scan and seems reasonable (adding patterns to fuse into the op can be done as follow up).

File: tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir]: Done (note that it was done to be similar to other surrounding tests, I can change it back for coherence if needed).

File: tensorflow/compiler/mlir/lite/tests/ops.mlir
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/tests/ops.mlir]: Done

File: tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir]: From my understanding it was done to make it easier to have a single return tensor without having to mess with shapes. It's done in a similar way for 2D average and max pooling tests.

**Comment by Tessil** [comment]
@gbaned @jpienaar Thanks for the review, I addressed the comments and updated the PR.

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned Done, thanks

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @sirakiin Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned @sirakiin Would it be possible to check what is blocking this PR internally as it has been open for 18 months and approved two times before? We can eventually close this PR and leave it internally to you?

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? It will help us to check the internal errors. Thank you!

**Review by alankelly — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil  Any update on this PR? Please. Thank you!

**Comment by Tessil** [comment]
Thanks I fixed the conflicts.

Though with the new 'BuiltinOptions2`, we have some problems with the generated `tensorflow/compiler/mlir/lite/operator_converters.inc` I'm not sure how to solve. @alankelly could you eventually take a look when you have some time? Thanks!

**Comment by gbaned** [comment]
Hi @Tessil Still, conflicts are appearing. Can you please resolve those? Thank you!

**Comment by gbaned** [comment]
Hi @alankelly Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @Tessil Can you please resolve conflicts? Thank you!

**Comment by damiles** [comment]
> Hi @Tessil Can you please resolve conflicts? Thank you!

I did a MR to @Tessil (https://github.com/Tessil/tensorflow/pull/1) to solve the conflict, but looks like he is not receiving notifications :'(

**Comment by gbaned** [comment]
Hi @Tessil I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! Thank you for your contribution.

**Comment by Tessil** [comment]
@gbaned Sorry for the late reply, we can leave the PR as closed for now. It has been open for more than two years and we don't have enough time to maintain it currently.

**Comment by damiles** [comment]
If it's possible, I want to maintain it. This feature is needed for some projects that I'm working on.

**Comment by Tessil** [comment]
@damiles Sorry for the delay. Feel free to open a new PR using the code of this one as long as the commit history of this PR is preserved. It might be good to check the interest of the TensorFlow team in this feature beforehand though.

**Pull request closed without merging**
